pkg/githubreconciler/issuemanager/session.go

```diff
@@ -0,0 +1,301 @@
+			}
+
+			if !needsUpdate {
+				log.Infof("Issue #%d is up to date, no refresh needed", existingIssue.GetNumber())
+				issueURLs[i] = existingIssue.GetHTMLURL()
+				continue
+			}
+
+			// Update existing issue
+			url, err := s.updateIssue(ctx, existingIssue, data, allLabels)
+			if err != nil {
+				return nil, fmt.Errorf("updating issue #%d: %w", existingIssue.GetNumber(), err)
+			}
+			issueURLs[i] = url
+		} else {
+			// Create new issue
+			url, err := s.createIssue(ctx, data, allLabels)
+			if err != nil {
+				return nil, fmt.Errorf("creating issue: %w", err)
+			}
+			issueURLs[i] = url
+		}
+	}
+
+	return issueURLs, nil
+}
+
+// CloseAnyOutstanding closes any existing issues that don't match the desired data set.
+// It uses the matcher function to determine which issues should be kept.
+// If message is non-empty, it posts the message as a comment before closing.
+func (s *IssueSession[T]) CloseAnyOutstanding(
+	ctx context.Context,
+	desired []*T,
+	matcher func(T, T) bool,
+	message string,
+) error {
+	log := clog.FromContext(ctx)
+
+	for _, issue := range s.existingIssues {
+		// Extract embedded data from issue
+		existing, err := s.manager.extractData(issue.GetBody())
+		if err != nil {
+			log.Warnf("Failed to extract data from issue #%d body, skipping: %v", issue.GetNumber(), err)
+			continue
+		}
+
+		// Check if this issue matches any desired data
+		matched := false
+		for _, data := range desired {
+			if matcher(*existing, *data) {
+				matched = true
+				break
+			}
+		}
+
+		if !matched {
+			// Close the issue
+			log.Infof("Closing issue #%d as it's not in the desired set", issue.GetNumber())
+
+			// Post message as a comment if provided
+			if message != "" {
+				if _, _, err := s.client.Issues.CreateComment(ctx, s.resource.Owner, s.resource.Repo, issue.GetNumber(), &github.IssueComment{
+					Body: github.Ptr(message),
+				}); err != nil {
+					return fmt.Errorf("posting comment on issue #%d: %w", issue.GetNumber(), err)
+				}
+			}
+
+			// Close the issue
+			if _, _, err := s.client.Issues.Edit(ctx, s.resource.Owner, s.resource.Repo, issue.GetNumber(), &github.IssueRequest{
+				State: github.Ptr("closed"),
+			}); err != nil {
+				return fmt.Errorf("closing issue #%d: %w", issue.GetNumber(), err)
+			}
+
+			log.Infof("Closed issue #%d", issue.GetNumber())
+		}
+	}
+
+	return nil
+}
+
+// findMatchingIssue finds an existing issue that matches the given data using the matcher function.
+// Returns nil if no match is found.
+func (s *IssueSession[T]) findMatchingIssue(ctx context.Context, data *T, matcher func(T, T) bool) *github.Issue {
+	log := clog.FromContext(ctx)
+
+	for _, issue := range s.existingIssues {
+		existing, err := s.manager.extractData(issue.GetBody())
+		if err != nil {
+			log.Warnf("Failed to extract data from issue #%d body, skipping: %v", issue.GetNumber(), err)
+			continue
+		}
+
+		if matcher(*existing, *data) {
+			return issue
+		}
```

Comment:
I would eliminate the `matcher` and replace it with an interface constraint on `*T`.

Really what the `matcher` is doing is saying "Are these two *T equivalent?" and I think I'd rather leverage generic constraints for that then plumb through a function.